telnet/src/lib.rs

```diff
@@ -54,33 +50,60 @@ impl TelnetProvider {
     }
 
     fn configure(&self, config: CapabilityConfiguration) -> MessageHandlerResult {
+        if let Some(listener) = self.listeners.read().unwrap().get(&config.module) {
+            debug!(
+                "Telnet session for actor {} already listening on {}",
+                listener.local_addr().unwrap(),
+                &config.module
+            );
+            return Ok(vec![]);
+        }
+
         session::start_server(
-            std::fs::read_to_string(&config.values.get("MOTD").unwrap_or(&"".to_string()))?,
+            config
+                .values
+                .get("MOTD")
+                .map_or_else(|| "".to_string(), |motd| motd.to_string()),
             config
                 .values
                 .get("PORT")
-                .unwrap_or(&"3000".to_string())
-                .parse()
-                .unwrap(),
+                .map_or_else(|| Ok(3000), |p| p.parse())?,
             &config.module,
             self.dispatcher.clone(),
             self.outbounds.clone(),
+            self.listeners.clone(),
         );
-
         Ok(vec![])
     }
 
-    fn deconfigure(&self, _config: CapabilityConfiguration) -> MessageHandlerResult {
-        // Handle removal of resources claimed by an actor here
-        // TODO: terminate the telnet server for this actor
+    fn deconfigure(&self, config: CapabilityConfiguration) -> MessageHandlerResult {
+        debug!("Shutting down telnet session for actor {}", &config.module);
+        // Remove actor session, shutdown TCP listener
+        if self
+            .listeners
+            .write()
+            .unwrap()
+            .remove(&config.module)
+            .is_none()
+        {
+            warn!(
+                "Attempted to deconfigure actor {}, but it was not configured",
+                &config.module
+            );
+        }
         Ok(vec![])
     }
 
     /// Sends a text message to the appropriate socket
-    fn send_text(&self, _actor: &str, msg: TelnetMessage) -> MessageHandlerResult {
-        let outbound = self.outbounds.read().unwrap()[&msg.session].clone();
-        outbound.send(msg.text).unwrap();
-        Ok(vec![])
+    fn send_text(
+        &self,
+        _actor: &str,
+        msg: SendTextArgs,
+    ) -> Result<Vec<u8>, Box<dyn Error + Sync + Send>> {
+        match self.outbounds.read().unwrap().get(&msg.session).clone() {
+            Some(outbound) => serialize(outbound.send(msg.text).is_ok()),
```

Comment:
Are we really serializing a boolean as the result here? If that's the contract in the widl then I think it needs to be changed because right now the message pack serialization requires a wrapper object, e.g. we can't return scalars from functions.

Reply:
It does look like we are returning `bool`s here: https://github.com/wasmCloud/actor-interfaces/blob/main/telnet/telnet.widl#L4-L6. I wonder if this was an artifact of an older widl design.

FWIW, the send text operation does work, but I haven't tried to actually unwrap the result and use it. Should returning a scalar with msgpack cause a panic right away or when you try and access the value? A result type would be more useful in general in order to pass error messages / failures

Reply:
If we don't use the result of that operation, then that's fine - it won't matter what we use because we never attempt to deserialize it.

However, if we ever attempt to deserialize a "floating bool" that has no wrapping container, that bool won't be deserializable with messagepack as coded in the actor interface libraries. The same code that wraps the function types in a unary wrapper for input should be applied to output as well, automatically creating a wrapper "args"... but that's an enhancement for another day.

Reply:
The error will only occur if the other side of a conversation attempts to de-serialize the scalar via messagepack.

Reply:
Before merging this, should I go ahead and address this before we go further? I saw the approval but I'm happy to at least change the result type in the WIDL and update it here.

Reply:
Yeah we can just update the widl to indicate a void. It won't change the `Vec<u8>` return type of the capability provider since all operations return that.

Reply:
As long as I remember correctly too, updating the return type to `void` just ends up `map`ping the return type to `()`, which is pretty neat since it won't affect here even if we do return a bool.

Reply:
Opened up https://github.com/wasmCloud/actor-interfaces/pull/75 to update the interface, will upgrade this provider to that interface version `0.2.0` once it's published. For now, going to go ahead and merge this.